File: 337.打家劫舍III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        // 递归超时
        // if(!root)
        //     return 0;
        // int val = root->val;
        // if(root->left)
        //     val += rob(root->left->left) + rob(root->left->right);
        // if(root->right)
        //     val += rob(root->right->left) + rob(root->right->right);
        // return max(val, rob(root->left) +rob(root->right));

        // 遍历每一颗子树，每颗子树返回两个值
        // 分别是包含根节点的最大值和不包含根节点的最大值
        auto res = helper(root);
        return max(res[0], res[1]);
    }

    vector<int> helper(TreeNode* root){
        if(!root)
            return {0, 0};
        auto left = helper(root->left);
        auto right = helper(root->right);
        return {max(left[0], left[1]) + max(right[0], right[1])}, root->val + left[0] + right[0]};
    }
};

File: 树/337.打家劫舍III.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 let map = new Map();
 var rob = function (root) {
 
     // 0表示选中的最大值，1表示未选中的最大值
     // const dfs = (root) => {
     //     if (!root)
     //         return [0, 0];
     //     let l = dfs(root.left);
     //     let r = dfs(root.right);
     //     let selected = root.val + l[1] + r[1];
     //     let noSelected = Math.max(l[0], l[1]) + Math.max(r[0], r[1]);
     //     return [selected, noSelected];
     // }
 
     // let val = dfs(root);
     // return Math.max(val[0], val[1]);
 
     // 单纯递归超时，记录已经访问的值就不超时
     if (!root)
         return 0;
     if (map.has(root))
         return map.get(root);
     let val = root.val;
     if (root.left)
         val += rob(root.left.left) + rob(root.left.right);
     if (root.right)
         val += rob(root.right.left) + rob(root.right.right);
     let res = Math.max(val, rob(root.left) + rob(root.right));
     map.set(root, res);
     return res;
 };
 
 